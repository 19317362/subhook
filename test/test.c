#include <stdio.h>
#include <subhook.h>

subhook_t hfoo, hbar;

void foo() {
	printf("foo() called\n");
}

void foo_hook() {
	printf("foo_hook() called\n");
}

void bar() {
	printf("bar() called\n");
}

void bar_hook() {
	printf("bar_hook() called\n");
	subhook_remove(hbar);
	bar();
	subhook_install(hbar);
}

int main(int argc, char **argv) {
	hfoo = subhook_new();
	subhook_set_src(hfoo, (void*)foo);
	subhook_set_dst(hfoo, (void*)foo_hook);
	foo();
	subhook_install(hfoo);
	foo();
	subhook_remove(hfoo);
	foo();
	subhook_free(hfoo);

	hbar = subhook_new();
	subhook_set_src(hbar, (void*)bar);
	subhook_set_dst(hbar, (void*)bar_hook);
	bar();
	subhook_install(hbar);
	bar();
	subhook_remove(hbar);
	bar();
	subhook_free(hbar);
}
