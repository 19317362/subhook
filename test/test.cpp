#include <cstdio>
#include <subhook.h>

typedef void (*foo_func_t)();

foo_func_t foo_tr = 0;

void foo() {
  std::printf("foo() called\n");
}

void foo_hooked() {
  std::printf("foo_hooked() called\n");
}

void foo_hooked_tr() {
  std::printf("foo_hooked_tr() called\n");
  foo_tr();
}

int main() {
  std::printf("Testing initial install\n");

  subhook::Hook foo_hook;
  foo_hook.Install((void *)foo,
                   (void *)foo_hooked,
                   subhook::HookOption64BitOffset);
  foo();
  foo_hook.Remove();
  foo();

  std::printf("Testing re-install\n");

  foo_hook.Install();
  foo();
  foo_hook.Remove();
  foo();

  std::printf("Testing trampoline\n");

  subhook::Hook foo_hook_tr((void *)foo,
                            (void *)foo_hooked_tr,
                            subhook::HookOption64BitOffset);
  foo_hook_tr.Install();
  foo_tr = (foo_func_t)foo_hook_tr.GetTrampoline();
  foo();
}
