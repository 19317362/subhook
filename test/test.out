Testing initial install
foo_hooked\(\) called
foo\(\) called
Testing re-install
foo_hooked\(\) called
foo\(\) called
Testing trampoline
foo\(\) called
